connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java

```diff
@@ -0,0 +1,578 @@
+        return new MultiFieldPaths(Arrays.asList(paths));
+    }
+
+    public static MultiFieldPaths of(List<SingleFieldPath> paths) {
+        return new MultiFieldPaths(paths);
+    }
+
+    public static MultiFieldPaths of(Set<String> fields, FieldSyntaxVersion syntaxVersion) {
+        return new MultiFieldPaths(fields.stream()
+                .map(f -> SingleFieldPath.of(f, syntaxVersion))
+                .collect(Collectors.toList()));
+    }
+
+    public static MultiFieldPaths of(List<String> fields, FieldSyntaxVersion syntaxVersion) {
+        return new MultiFieldPaths(fields.stream()
+                .map(f -> SingleFieldPath.of(f, syntaxVersion))
+                .collect(Collectors.toList()));
+    }
+
+    Map<String, Object> buildPathTree(List<SingleFieldPath> paths, int stepIdx, Map<String, Object> pathTree) {
+        if (paths.size() == 1) { // optimize for paths with a single member
+            SingleFieldPath path = paths.get(0);
+            if (path != null) {
+                if (path.stepAt(stepIdx + 1) == null) { // if last path step
+                    pathTree.put(path.stepAt(stepIdx), path);
+                } else {
+                    pathTree.put(path.stepAt(stepIdx),
+                            buildPathTree(paths, stepIdx + 1, new HashMap<>()));
+                }
+            }
+        } else {
+            // group paths by prefix
+            final Map<String, List<SingleFieldPath>> groups = new HashMap<>();
+            for (SingleFieldPath path : paths) {
+                if (path != null) {
+                    String step = path.stepAt(stepIdx);
+                    if (step != null) {
+                        groups.computeIfPresent(step, (s, fieldPaths) -> {
+                            for (SingleFieldPath other : fieldPaths) {
+                                // avoid overlapping paths
+                                if (!path.equals(other)
+                                        && (other.stepAt(stepIdx + 1) == null
+                                        || path.stepAt(stepIdx + 1) == null)) {
+                                    throw new IllegalArgumentException(
+                                            "Path " + other + " and " + path + " are overlapping. "
+                                                    + "Paths need to point to leaf values");
+                                }
+                            }
+                            if (!fieldPaths.contains(path)) {
+                                fieldPaths.add(path);
+                            }
+                            return fieldPaths;
+                        });
+                        groups.computeIfAbsent(step, s -> {
+                            List<SingleFieldPath> fieldPaths = new ArrayList<>();
+                            fieldPaths.add(path);
+                            return fieldPaths;
+                        });
+                    }
+                }
+            }
```

Comment:
I think this can be simplified using the Java 8 streams API:

```suggestion
            final Map<String, List<SingleFieldPath>> groups = paths.stream()
                    .filter(p -> p.stepAt(stepIdx) != null)
                    .collect(Collectors.groupingBy(
                            path -> path.stepAt(stepIdx)
                    ));
```

If we really need to check to make sure we don't add the same path twice to a value in that map, we can collect into a `Set` instead:
```java
            final Map<String, Set<SingleFieldPath>> groups = paths.stream()
                    .filter(p -> p.stepAt(stepIdx) != null)
                    .collect(Collectors.groupingBy(
                            path -> path.stepAt(stepIdx),
                            Collectors.toSet()
                    ));
```

If we really need to do the check for overlapping paths, we can do that after creating the `groups` field by iterating over each of its values.

Reply:
I like this! I was trying to be cautious of users adding overlapping paths. But by using Sets we could conflate them and only use the children paths.
Giving a try now.